pkg/datapath/linux/ipsec/probe_linux.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// +build linux
+
+package ipsec
+
+import (
+	"encoding/hex"
+	"net"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/linux_defaults"
+
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	dummyIP  = "169.254.169.254"
+	aeadKey  = "4242424242424242424242424242424242424242"
+	aeadAlgo = "rfc4106(gcm(aes))"
+	stateId  = 42
+)
+
+func initDummyXfrmState() *netlink.XfrmState {
+	state := ipSecNewState()
+
+	k, _ := hex.DecodeString(aeadKey)
+	state.Aead = &netlink.XfrmStateAlgo{
+		Name:   aeadAlgo,
+		Key:    k,
+		ICVLen: 128,
+	}
+	state.Spi = int(stateId)
+	state.Reqid = stateId
+
+	state.Src = net.ParseIP(dummyIP)
+	state.Dst = net.ParseIP(dummyIP)
+	return state
+}
+
+func createDummyXfrmState(state *netlink.XfrmState) error {
+	state.Mark = &netlink.XfrmMark{
+		Value: linux_defaults.RouteMarkDecrypt,
+		Mask:  linux_defaults.IPsecMarkMaskIn,
+	}
+	state.OutputMark = &netlink.XfrmMark{
+		Value: linux_defaults.RouteMarkDecrypt,
+		Mask:  linux_defaults.RouteMarkMask,
+	}
+	return netlink.XfrmStateAdd(state)
+}
+
+func ProbeXfrmStateOutputMask() (bool, error) {
+	state := initDummyXfrmState()
+	err := createDummyXfrmState(state)
+	if err != nil {
+		return false, err
+	}
+	defer netlink.XfrmStateDel(state)
+
+	probedState, err := netlink.XfrmStateGet(state)
+	if err != nil || probedState == nil || probedState.OutputMark == nil {
+		return false, err
+	}
```

Comment:
The caller seems to be avoiding a standard pattern of `if ..., err := foo(); err != nil` because of the `probedState == nil || probedState.OutputMark == nil` cases here, could we detect those specific cases and just set `err = errors.New("ipsec output mark attribute missing from xfrm probe")` (or some better-worded message)? Then the caller doesn't have to additionally check whether the lack of `err` indicates a lack of support.

Reply:
Makes sense. Changed.